crates/sui-core/src/transaction_input_loader.rs

```diff
@@ -0,0 +1,355 @@
+        }
+
+        Ok((
+            results
+                .into_iter()
+                .map(Option::unwrap)
+                .collect::<Vec<_>>()
+                .into(),
+            receiving_results.into(),
+        ))
+    }
+
+    /// Reads input objects assuming a synchronous context such as the end of epoch transaction.
+    /// By "synchronous" we mean that it is safe to read the latest version of all shared objects,
+    /// as opposed to relying on the shared input version assignment.
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_synchronous_execution(
+        &self,
+        tx_digest: &TransactionDigest,
+        input_object_kinds: &[InputObjectKind],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<InputObjects> {
+        self.read_objects_for_synchronous_execution_impl(
+            Some(tx_digest),
+            input_object_kinds,
+            &[],
+            protocol_config,
+        )
+        .await
+        .map(|r| r.0)
+    }
+
+    /// Read input objects for a dry run execution.
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_dry_run_exec(
+        &self,
+        tx_digest: &TransactionDigest,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<(InputObjects, ReceivingObjects)> {
+        self.read_objects_for_synchronous_execution_impl(
+            Some(tx_digest),
+            input_object_kinds,
+            receiving_objects,
+            protocol_config,
+        )
+        .await
+    }
+
+    /// Read input objects for dev inspect
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_dev_inspect(
+        &self,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+        protocol_config: &ProtocolConfig,
+    ) -> SuiResult<(InputObjects, ReceivingObjects)> {
+        self.read_objects_for_synchronous_execution_impl(
+            None,
+            input_object_kinds,
+            receiving_objects,
+            protocol_config,
+        )
+        .await
+    }
+
+    /// Read the inputs for a transaction that is ready to be executed.
+    ///
+    /// shared_lock_store is used to resolve the versions of any shared input objects.
+    ///
+    /// This function panics if any inputs are not available, as TransactionManager should already
+    /// have verified that the transaction is ready to be executed.
+    ///
+    /// The tx_digest is provided here to support the following optimization (not yet implemented):
+    /// All the owned input objects will likely have been loaded during transaction signing, and
+    /// can be stored as a group with the transaction_digest as the key, allowing the lookup to
+    /// proceed with only a single hash map lookup. (additional lookups may be necessary for shared
```

Comment:
well, it could actually still be done, although we would have to invalidate the cache when objects are received - probably more trouble than its worth most of the time